backend/server/query_analysis.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+
+import json
+from typing import Any, Dict, List, Optional, Sequence, Union
+
+from fastapi import HTTPException
+from jsonschema import Draft202012Validator, exceptions
+from langchain_core.messages import AIMessage, AnyMessage
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from langchain_core.prompts.chat import MessageLikeRepresentation
+from langchain_core.runnables import chain
+from langserve import CustomUserType
+from pydantic import BaseModel, Field, validator
+from typing_extensions import TypedDict
+
+from db.models import QueryAnalysisExample as DBQueryAnalysisExample
+from db.models import QueryAnalyzer
+from extraction.utils import convert_json_schema_to_openai_schema
+from server.settings import get_model
+from server.validators import validate_json_schema
+
+# Instantiate the model
+model = get_model()
+
+
+class QueryAnalysisExample(BaseModel):
+    """An example query analysis.
+
+    This example consists of input messages and the expected queries.
+    """
+
+    messages: List[AnyMessage] = Field(..., description="The input messages")
+    output: List[Dict[str, Any]] = Field(
+        ..., description="The expected output of the example. A list of objects."
+    )
+
+
+class QueryAnalysisRequest(CustomUserType):
+    """Request body for the query analyzer endpoint."""
+
+    messages: List[AnyMessage] = Field(
+        ..., description="The messages to generates queries from."
+    )
+    json_schema: Union[Dict[str, Any], List[Dict[str, Any]]] = Field(
+        ...,
+        description="JSON schema(s) that describes what a query looks like",
+        alias="schema",
+    )
+    instructions: Optional[str] = Field(
+        None, description="Supplemental system instructions."
+    )
+    examples: Optional[List[QueryAnalysisExample]] = Field(
+        None, description="Examples of optimized queries."
+    )
+
+    @validator("json_schema")
+    def validate_schema(cls, v: Any) -> Dict[str, Any]:
+        """Validate the schema."""
+        # validate_json_schema(v)
```

Comment:
@eyurtsev what's this needed for and do you know why it fails when `data` array items are an `anyOf`?

Reply:
but code works

Reply:
oh oh i see